falcon/cluster/cluster.py

```diff
@@ -920,3 +746,76 @@ def _get_cluster_medoid_index(
         if row_avg < min_avg:
             min_i, min_avg = row_i, row_avg
     return cluster_mask[min_i]
+
+
+def compute_condensed_distance_matrix(
+    spec_tuples: List[similarity.SpectrumTuple],
+    fragment_mz_tol: float,
+    allow_shift: bool,
+) -> np.ndarray:
+    """
+    Compute the condensed pairwise distance matrix for the given spectra.
+
+    Parameters
+    ----------
+    spec_tuples : List[similarity.SpectrumTuple]
+        The spectra to compute the pairwise distance matrix for.
+    fragment_mz_tolerance : float
+        The fragment m/z tolerance.
+    allow_shift : bool
+        Whether to allow peak shifts.
+
+    Returns
+    -------
+    np.ndarray
+        The condensed pairwise distance matrix.
+    """
+    n = len(spec_tuples)
+    condensed_dist_matrix = np.zeros(n * (n - 1) // 2)
+
+    def worker(i, j):
+        spec_tup1 = spec_tuples[i]
+        spec_tup2 = spec_tuples[j]
+        sim, _, _, _, _, _, _ = similarity._cosine_fast(
+            spec_tup1, spec_tup2, fragment_mz_tol, allow_shift
+        )
+        distance = 1.0 - sim
+        idx = condensed_index(i, j, n)
+        condensed_dist_matrix[idx] = abs(distance)
```

Comment:
**question:** Not sure I understand this. Why is this necessary? The distance should always be >= 0, so what is this supposed to do?

Reply:
Previously, I ran into the issue of the distance being slightly less than 0 for some reason, but that doesn't seem to be the case anymore since using `_cosine_fast`

Reply:
Ok. For future reference, using a "clip" to 0 behavior is better in that case. The absolute value will be non-zero positive, but presumably those negative values are due to some rounding errors and should be just 0 instead.

Reply:
It's still a problem on our toy dataset. Some similarities are just slightly > 1. I'll clip in `_cosine_fast`.